Decode JPEG XL images and PDF /JXLDecode streams with jxldec.

The decoder adapter lives in src/mupdf/load-jxl.cpp and load-jxl.h. Reuse the self-describing image path for PDF colorspaces,
Decode arrays and masks. Preserve JXL filters when expanding PDF streams.

---
diff --git a/include/mupdf/fitz/compressed-buffer.h b/include/mupdf/fitz/compressed-buffer.h
index 296079618..103ed46a0 100644
--- a/include/mupdf/fitz/compressed-buffer.h
+++ b/include/mupdf/fitz/compressed-buffer.h
@@ -177,6 +177,7 @@ enum
 	FZ_IMAGE_TIFF,
 	FZ_IMAGE_PSD,
 	FZ_IMAGE_WEBP,
+	FZ_IMAGE_JXL,
 };
 
 /**
diff --git a/include/mupdf/pdf/name-table.h b/include/mupdf/pdf/name-table.h
index 1a5586635..873f92631 100644
--- a/include/mupdf/pdf/name-table.h
+++ b/include/mupdf/pdf/name-table.h
@@ -314,6 +314,7 @@ PDF_MAKE_NAME("JBIG2Decode", JBIG2Decode)
 PDF_MAKE_NAME("JBIG2Globals", JBIG2Globals)
 PDF_MAKE_NAME("JPXDecode", JPXDecode)
 PDF_MAKE_NAME("JS", JS)
+PDF_MAKE_NAME("JXLDecode", JXLDecode)
 PDF_MAKE_NAME("JavaScript", JavaScript)
 PDF_MAKE_NAME("K", K)
 PDF_MAKE_NAME("Keywords", Keywords)
diff --git a/source/fitz/image.c b/source/fitz/image.c
index 0ab603f8b..1fe500a74 100644
--- a/source/fitz/image.c
+++ b/source/fitz/image.c
@@ -24,6 +24,7 @@
 
 #include "context-imp.h"
 #include "image-imp.h"
+#include "load-jxl.h"
 #include "pixmap-imp.h"
 
 #include <string.h>
@@ -791,6 +792,7 @@ fz_is_lossy_image(fz_context *ctx, fz_image *image)
 		case FZ_IMAGE_JPEG:
 		case FZ_IMAGE_JPX:
 		case FZ_IMAGE_JXR:
+		case FZ_IMAGE_JXL:
 			return 1;
 		}
 	}
@@ -842,8 +844,12 @@ compressed_image_get_pixmap(fz_context *ctx, fz_image *image_, fz_irect *subarea
 	case FZ_IMAGE_JXR:
 		tile = fz_load_jxr(ctx, image->buffer->buffer->data, image->buffer->buffer->len);
 		break;
+	case FZ_IMAGE_JXL:
 	case FZ_IMAGE_JPX:
-		tile = fz_load_jpx(ctx, image->buffer->buffer->data, image->buffer->buffer->len, image->super.colorspace);
+		if (image->buffer->params.type == FZ_IMAGE_JXL)
+			tile = fz_load_jxl(ctx, image->buffer->buffer->data, image->buffer->buffer->len, image->super.colorspace);
+		else
+			tile = fz_load_jpx(ctx, image->buffer->buffer->data, image->buffer->buffer->len, image->super.colorspace);
 		if (image->super.use_colorkey)
 		{
 			size_t len;
@@ -1436,6 +1442,7 @@ fz_image_type_name(int type)
 	case FZ_IMAGE_PNM: return "pnm";
 	case FZ_IMAGE_TIFF: return "tiff";
 	case FZ_IMAGE_WEBP: return "webp";
+	case FZ_IMAGE_JXL: return "jxl";
 	}
 }
 
@@ -1459,6 +1466,7 @@ fz_lookup_image_type(const char *type)
 	if (!strcmp(type, "pnm")) return FZ_IMAGE_PNM;
 	if (!strcmp(type, "tiff")) return FZ_IMAGE_TIFF;
 	if (!strcmp(type, "webp")) return FZ_IMAGE_WEBP;
+	if (!strcmp(type, "jxl")) return FZ_IMAGE_JXL;
 	return FZ_IMAGE_UNKNOWN;
 }
 
@@ -1499,6 +1507,9 @@ fz_recognize_image_format(fz_context *ctx, unsigned char p[12])
 	if (p[0] == 'R' && p[1] == 'I' && p[2] == 'F' && p[3] == 'F' &&
 		p[8] == 'W' && p[9] == 'E' && p[10] == 'B' && p[11] == 'P')
 		return FZ_IMAGE_WEBP;
+	if ((p[0] == 0xff && p[1] == 0x0a) ||
+		!memcmp(p, "\0\0\0\x0cJXL \r\n\x87\n", 12))
+		return FZ_IMAGE_JXL;
 	return FZ_IMAGE_UNKNOWN;
 }
 
@@ -1560,6 +1571,9 @@ fz_new_image_from_buffer(fz_context *ctx, fz_buffer *buffer)
 		fz_load_jbig2_info(ctx, buf, len, &w, &h, &xres, &yres, &cspace);
 		bpc = 1;
 		break;
+	case FZ_IMAGE_JXL:
+		fz_load_jxl_info(ctx, buf, len, &w, &h, &xres, &yres, &cspace);
+		break;
 	case FZ_IMAGE_WEBP:
 		fz_load_webp_info(ctx, buf, len, &w, &h, &xres, &yres, &cspace);
 		break;
diff --git a/source/pdf/pdf-image.c b/source/pdf/pdf-image.c
index 5c2a6a30f..d7998b00d 100644
--- a/source/pdf/pdf-image.c
+++ b/source/pdf/pdf-image.c
@@ -23,10 +23,12 @@
 #include "mupdf/fitz.h"
 #include "mupdf/pdf.h"
 
+#include "load-jxl.h"
+
 #include <string.h>
 
-static fz_image *pdf_load_jpx_as_compressed_image(fz_context *ctx, pdf_document *doc, pdf_obj *dict);
-static fz_image *pdf_load_jpx_as_compressed_image_mask(fz_context *ctx, pdf_document *doc, pdf_obj *dict);
+static fz_image *pdf_load_full_image(fz_context *ctx, pdf_document *doc, pdf_obj *dict, int type);
+static fz_image *pdf_load_full_image_mask(fz_context *ctx, pdf_document *doc, pdf_obj *dict, int type);
 
 static fz_image *
 pdf_load_image_imp(fz_context *ctx, pdf_document *doc, pdf_resource_stack *rdb, pdf_obj *dict, fz_stream *cstm, int forcemask)
@@ -50,14 +52,21 @@ pdf_load_image_imp(fz_context *ctx, pdf_document *doc, pdf_resource_stack *rdb,
 	int i;
 	fz_compressed_buffer *buffer;
 
-	/* special case for JPEG2000 which ignore the normal colorspace and image parameters */
-	if (pdf_is_jpx_image(ctx, dict))
+	/* JPEG2000 and JPEG XL carry their own image parameters. */
+	int type = FZ_IMAGE_JPX;
+	pdf_obj *filter = pdf_dict_get(ctx, dict, PDF_NAME(Filter));
+	int is_jxl = pdf_name_eq(ctx, filter, PDF_NAME(JXLDecode));
+	for (i = 0; i < pdf_array_len(ctx, filter); i++)
+		is_jxl |= pdf_name_eq(ctx, pdf_array_get(ctx, filter, i), PDF_NAME(JXLDecode));
+	if (is_jxl)
+		type = FZ_IMAGE_JXL;
+	if (is_jxl || pdf_is_jpx_image(ctx, dict))
 	{
 		if (cstm)
-			fz_throw(ctx, FZ_ERROR_SYNTAX, "inline JPXDecode image");
+			fz_throw(ctx, FZ_ERROR_SYNTAX, "inline JPXDecode/JXLDecode image");
 		if (forcemask)
-			return pdf_load_jpx_as_compressed_image_mask(ctx, doc, dict);
-		return pdf_load_jpx_as_compressed_image(ctx, doc, dict);
+			return pdf_load_full_image_mask(ctx, doc, dict, type);
+		return pdf_load_full_image(ctx, doc, dict, type);
 	}
 
 	w = pdf_to_int(ctx, pdf_dict_geta(ctx, dict, PDF_NAME(Width), PDF_NAME(W)));
@@ -247,7 +256,7 @@ pdf_is_jpx_image(fz_context *ctx, pdf_obj *dict)
 }
 
 static fz_image *
-pdf_load_jpx_as_compressed_image_mask(fz_context *ctx, pdf_document *doc, pdf_obj *dict)
+pdf_load_full_image_mask(fz_context *ctx, pdf_document *doc, pdf_obj *dict, int type)
 {
 	fz_buffer *buf = NULL;
 	fz_image *img = NULL;
@@ -259,7 +268,8 @@ pdf_load_jpx_as_compressed_image_mask(fz_context *ctx, pdf_document *doc, pdf_ob
 	buf = pdf_load_image_stream(ctx, doc, pdf_to_num(ctx, dict), &params, NULL, 0);
 	fz_try(ctx)
 	{
-		img = fz_new_jpx_image_from_buffer(ctx, buf, fz_device_gray(ctx));
+		img = type == FZ_IMAGE_JXL ? fz_new_jxl_image(ctx, buf, fz_device_gray(ctx)) :
+			fz_new_jpx_image_from_buffer(ctx, buf, fz_device_gray(ctx));
 	}
 	fz_always(ctx)
 	{
@@ -277,7 +287,7 @@ pdf_load_jpx_as_compressed_image_mask(fz_context *ctx, pdf_document *doc, pdf_ob
 }
 
 static fz_image *
-pdf_load_jpx_as_compressed_image(fz_context *ctx, pdf_document *doc, pdf_obj *dict)
+pdf_load_full_image(fz_context *ctx, pdf_document *doc, pdf_obj *dict, int type)
 {
 	fz_buffer *buf = NULL;
 	fz_image *img = NULL;
@@ -296,7 +306,7 @@ pdf_load_jpx_as_compressed_image(fz_context *ctx, pdf_document *doc, pdf_obj *di
 		if (obj)
 			cs = pdf_load_colorspace(ctx, obj);
 
-		img = fz_new_jpx_image_from_buffer(ctx, buf, cs);
+		img = type == FZ_IMAGE_JXL ? fz_new_jxl_image(ctx, buf, cs) : fz_new_jpx_image_from_buffer(ctx, buf, cs);
 		n = img->colorspace ? img->colorspace->n : 1;
 
 		obj = pdf_dict_geta(ctx, dict, PDF_NAME(Decode), PDF_NAME(D));
diff --git a/source/pdf/pdf-stream.c b/source/pdf/pdf-stream.c
index 2da7c7d2b..752aa0e98 100644
--- a/source/pdf/pdf-stream.c
+++ b/source/pdf/pdf-stream.c
@@ -200,6 +200,10 @@ build_compression_params(fz_context *ctx, pdf_obj *f, pdf_obj *p, fz_compression
 				params->u.jbig2.globals = pdf_load_jbig2_globals(ctx, g);
 		}
 	}
+	else if (pdf_name_eq(ctx, f, PDF_NAME(JXLDecode)))
+	{
+		params->type = FZ_IMAGE_JXL;
+	}
 	else if (pdf_name_eq(ctx, f, PDF_NAME(JPXDecode)))
 	{
 		params->type = FZ_IMAGE_JPX;
@@ -225,7 +229,8 @@ build_filter(fz_context *ctx, fz_stream *chain, pdf_document *doc, pdf_obj *f, p
 			pdf_name_eq(ctx, f, PDF_NAME(DCTDecode)) ||
 			pdf_name_eq(ctx, f, PDF_NAME(DCT)) ||
 			pdf_name_eq(ctx, f, PDF_NAME(JBIG2Decode)) ||
-			pdf_name_eq(ctx, f, PDF_NAME(JPXDecode))))
+			pdf_name_eq(ctx, f, PDF_NAME(JPXDecode)) ||
+			pdf_name_eq(ctx, f, PDF_NAME(JXLDecode))))
 	{
 		fz_warn(ctx, "Can't open image only stream for non-image purposes");
 		return fz_open_memory(ctx, (unsigned char *)"", 0);
diff --git a/source/pdf/pdf-write.c b/source/pdf/pdf-write.c
index bf83a4274..d30362231 100644
--- a/source/pdf/pdf-write.c
+++ b/source/pdf/pdf-write.c
@@ -1132,7 +1132,7 @@ static int is_image_filter(pdf_obj *s)
 		s == PDF_NAME(DCTDecode) || s == PDF_NAME(DCT) ||
 		s == PDF_NAME(RunLengthDecode) || s == PDF_NAME(RL) ||
 		s == PDF_NAME(JBIG2Decode) ||
-		s == PDF_NAME(JPXDecode);
+		s == PDF_NAME(JPXDecode) || s == PDF_NAME(JXLDecode);
 }
 
 static int filter_implies_image(fz_context *ctx, pdf_obj *o)
@@ -1152,14 +1152,15 @@ static int filter_implies_image(fz_context *ctx, pdf_obj *o)
 
 static int is_jpx_filter(fz_context *ctx, pdf_obj *o)
 {
-	if (o == PDF_NAME(JPXDecode))
+	if (o == PDF_NAME(JPXDecode) || o == PDF_NAME(JXLDecode))
 		return 1;
 	if (pdf_is_array(ctx, o))
 	{
 		int i, len;
 		len = pdf_array_len(ctx, o);
 		for (i = 0; i < len; i++)
-			if (pdf_array_get(ctx, o, i) == PDF_NAME(JPXDecode))
+			if (pdf_array_get(ctx, o, i) == PDF_NAME(JPXDecode) ||
+			pdf_array_get(ctx, o, i) == PDF_NAME(JXLDecode))
 				return 1;
 	}
 	return 0;
